### ISSUE TITLE
Improve backend-only authentication

### ISSUE THREAD

**Issue by webdeb** [hasura/graphql-engine#5147]
Hi guys,
I would like to suggest to change the current `backend-only` authentication flow which reads as follows from the docs:

> Backend only
> If a permission is marked as backend_only, the mutation is accessible to the given role only if x-hasura-use-backend-only-permissions session variable exists and is set to true and request is made with x-hasura-admin-secret set if any auth is configured.

I am not comfortable with this for the reason that we have to send the admin-secret key which is basically the `root` of the system. Since we are sending the admin-secret, the normal user JWT authorization is not working/skipped.

My suggestion is to only use a special ENV variable `BACKEND_ONLY_KEY` and send it from the backend server as `x-hasura-backend-only-key` to only **authenticate** the sender as a valid backend. The actual authorization rules should be applied as usual through the `authorization` header / or the admin-secret (only if you really have to!!!)

**Comment by njaremko** [comment]
Seems more or less fine. A couple comments:

- How is this different from using your "admin secret" with the `metadata`, `pgdump` and `config` APIs disabled? Is the idea that you could leave some of those enabled, and keep the admin key available for some specific use case?
     - https://hasura.io/docs/1.0/graphql/manual/deployment/production-checklist.html#disable-apis
- Env var should be prefixed with `HASURA_`
- Hasura is already parsing and validating the authorization JWT, and giving you the session variables in the payload, so you should probably be passing back the session variables as headers instead of making hasura parse and validate the JWT again

**Comment by webdeb** [comment]
1. yes, I just think it should be possible to use the graphql api and all its features without the the need of the `admin secret`. 
2. sure ;)
3. Imagine a direct call to the backend (like a file upload) and then to hasura to create the file entry, there are no `session_variables` at this point.

Example:
`Frontend` -> `Backend` -> `HASURA`

**Comment by njaremko** [comment]
Answers to 2 and 3 sound fine.

For (1):
- Is you're vision for this "backend key" to be something akin to an admin secret, but without access to non-graphql APIs? (A key that has unrestricted access to the graphql API, but can't do anything on the `metadata`, `pgdump`, or `config` apis)

**Comment by webdeb** [comment]
Yes the admin secret is **authorized** to do anything. The backend key only **authenticates** the backend, without permissions at all.

> If so, would you still leave any of those APIs enabled on your production instance?

it depends, I think most of us will use some of the apis (in the "startup mode", _while moving fast and breaking things_, but when even the docs have a **disable apis** section, then those apis are not required to run for production use, so the usage of the `admin secret` could be avoided.

**Comment by njaremko** [comment]
The reason I ask is because with those three APIs disabled, the admin secret seems to become you're desired backend key (it can do anything on graphql api, but that's it).

Unless you only want this key to say "this query is coming from my backend" while relying on the authorization header for all authorization stuff. In which case, I can see where you're coming from, but it seems a bit niche, so I'd guess it would take a while to be prioritized by the people at Hasura.

**Comment by webdeb** [comment]
> it can do anything on graphql api, but that's it

that's it? its unlimited CRUD

**Comment by njaremko** [comment]
> that's it? its unlimited CRUD

Yes, which is less than it could do otherwise, hence the "that's it".

I think the current official suggestion is to do the checks in your backend before you call Hasura, or manually pass in the relevant headers to scope the access, but I can understand you wanting to pass that on to Hasura if the JWT is already there. Sounds like you might be able to get what you want with just some precedence rules. What do you think of this? (This might be what you suggested in a previous issue, can't remember)

1. If admin secret is present alone, you can do whatever
2. If admin secret is present with hasura headers, these take precedence over authorization
3. If admin secret is present with authorization header, the request is now a backend request, but all authorization comes from the header (I think this is different from the current behaviour, which would make this a breaking change, but maybe the people at Hasura would be okay with it...unlikely though)

Without the above change, we would need your proposed "backend key" with the following rules:
1. If just the backend key is provided, it's a "backend" request, but the "unauthorized" role is used (assuming it's configured)
1. If the backend key is present with hasura headers, they take precedence
2. Otherwise, use either the authorization header JWT, or pass everything along to the auth webhook (whichever is configured)

**Comment by Motoxpro** [comment]
@webdeb Why not just create a new role with exactly the permissions you want? A "backend" role?

**Comment by lakshyasharma14** [comment]
@webdeb I was also looking for something like a backend key, for calls from the nextjs server-side calls. I don't want to use the admin key to provide some limited access to the backend.

**Comment by o-az** [comment]
I want something like the admin secret key but for readonly. Surprised this doesn't exist. The closest is having multiple admin secret keys but that's only available for Cloud and self-hosted Enterprise.
